### PR TITLE
Add down host cache

### PR DESCRIPTION
**Pull request by raihan687**
Adds a process-wide down-host cache for connection attempts.

Hosts that time out or are unreachable during TCP connection establishment are temporarily marked as down. New connection iterators and retry cycles move these hosts to the end of the attempt order, while retaining them as fallbacks.

The cache expires entries after 10 minutes and does not cache connection-refused, DNS, listener, TLS, or authentication failures.

### CODE REVIEW

File: internal/common/caches.go [hunk 1]
```diff
@@ -0,0 +1,412 @@
+/*
+** Copyright (c) 2026 Oracle and/or its affiliates.
+**
```
Comment (ejannett) [inline internal/common/caches.go:3]: the PR #82 will be hopefully merged soon. this file is not needed.

File: internal/driver/network/naming/down_hosts_cache_test.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
+/*
+** Copyright (c) 2026 Oracle and/or its affiliates.
+**
```
Comment (ejannett) [inline internal/driver/network/naming/down_hosts_cache_test.go:3]: PR #82 will bring testing , this file is not needed.

File: internal/driver/network/naming/down_hosts_cache.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+var sharedDownHostCache = common.NewSafeTTLCache[struct{}](
+	downHostCacheSize,
+	downHostCacheTTL,
+)
+
+// MarkDownHost records a host that could not be reached. The cache only
+// influences connection order; it never removes a host from the attempt list.
+func MarkDownHost(host string) {
+	if host != "" {
+		sharedDownHostCache.Put(host, struct{}{})
```
Comment (ejannett) [inline internal/driver/network/naming/down_hosts_cache.go:51]: No need to allocate a structure each time. 
using sharedDownHostCache.Put(host, host) is just fine

File: internal/driver/network/naming/iterator.go [hunk 4]
```diff
@@ -108,13 +109,18 @@ type ConnectionIterator struct {
 // DNS resolution happens during iterator creation - hostnames are resolved to IPs,
 // and each IP becomes a separate connection attempt.
 func NewConnectionIterator(ctx context.Context, rootNode *Node, connCtx *ConnectionContext) *ConnectionIterator {
+	return newConnectionIterator(ctx, rootNode, connCtx, sharedDownHostCache)
```
Comment (ejannett) [inline internal/driver/network/naming/iterator.go:112]: what is the need of passing a global reference as parameter ?

File: internal/driver/network/naming/iterator.go [hunk 5]
```diff
@@ -597,3 +610,54 @@ func (ci *ConnectionIterator) roundRobin(addresses []Address) {
...
+				return false
+			}
+		}
+		return true
+	})
+}
+
+func (ci *ConnectionIterator) isDownHost(address Address) bool {
+	if ci.downHostCache == nil {
+		return false
```
Comment (ejannett) [inline internal/driver/network/naming/iterator.go:636]: this case do not exists

File: internal/common/caches.go (hunk 1 above)
Comment (raihan687) [reply internal/common/caches.go:3]: Sure will do.

File: internal/driver/network/naming/iterator.go (hunk 5 above)
Comment (raihan687) [reply internal/driver/network/naming/iterator.go:636]: Sure will remove

File: internal/driver/network/naming/iterator.go (hunk 4 above)
Comment (raihan687) [reply internal/driver/network/naming/iterator.go:112]: Sure I’ll access the shared cache directly

File: internal/driver/network/naming/down_hosts_cache_test.go (hunk 2 above)
Comment (raihan687) [reply internal/driver/network/naming/down_hosts_cache_test.go:3]: Ok will remove this as well.

File: internal/driver/network/session/nsprotocol.go [hunk 6]
```diff
@@ -170,6 +181,40 @@ func (ns *networkSession) transportConnect(ctx context.Context, address transpor
...
+	if errors.As(err, &timeoutCause) {
+		return true
+	}
+
+	var sqlErr oracleErrors.SQLError
+	if errors.As(err, &sqlErr) && sqlErr.ErrorCode() == string(oracleErrors.CtxTimeout) {
+		return true
+	}
+
+	if errors.Is(err, context.DeadlineExceeded) {
```
Comment (ashichau) [inline internal/driver/network/session/nsprotocol.go]: hi raihan, i think 
if errors.Is(err, context.DeadlineExceeded) {
    return true
}

return errors.As(err, &netErr) && netErr.Timeout()

context.DeadlineExceeded can mean the caller’s overall context expired—not that the host is down. For example, if the caller gives the entire connection attempt a 1-second deadline, that deadline may expire while trying a healthy host. The code would still cache that host as down for 10 minutes.
So, basically 
A caller-level deadline can expire before a host is actually proven unreachable. That healthy host would then be deprioritized for 10 minutes.
Consider caching only transport-specific connection timeouts, such as TransportConnectTimeout,

File: internal/common/caches.go (hunk 1 above)
Comment (ejannett) [reply internal/common/caches.go:3]: This file should be discarded right?

File: internal/driver/network/naming/down_hosts_cache.go (hunk 3 above)
Comment (raihan687) [reply internal/driver/network/naming/down_hosts_cache.go:51]: I kept struct{}{} because this cache only needs membership; the stored value is never read. An empty struct has zero size, so it does not allocate per entry. Put(key, key) would work too, but it stores an unused string value and therefore uses unnecessary memory.

File: internal/common/caches.go (hunk 1 above)
Comment (raihan687) [reply internal/common/caches.go:3]: yes

File: internal/driver/network/naming/down_hosts_cache_test.go (hunk 2 above)
Comment (raihan687) [reply internal/driver/network/naming/down_hosts_cache_test.go:3]: The Pr 82 tests covers generic cache APIs, while this file covers the down-host-cache functionality built on top of them. So I think we should keep this file.

Comment (ejannett) [reply internal/driver/network/naming/down_hosts_cache_test.go]: please rename that file to down_hosts_test.go

File: internal/driver/network/naming/iterator.go [hunk 7]
```diff
@@ -597,3 +604,55 @@ func (ci *ConnectionIterator) roundRobin(addresses []Address) {
...
 	addresses[len(addresses)-1] = first
 }
+
+// reorderAddressesByDownHostStatus puts recently unreachable hosts at the end
+// of an attempt cycle. Cached hosts are retained as fallbacks.
+func (ci *ConnectionIterator) reorderAddressesByDownHostStatus(addresses []Address) {
+	stablePartition(addresses, ci.isDownHost)
+}
+
+func (ci *ConnectionIterator) reorderDescriptionsByDownHostStatus(attempts []DescriptionAttempts) {
```
Comment (ejannett) [inline internal/driver/network/naming/iterator.go]: add missing doc

File: internal/driver/network/naming/iterator.go [hunk 8]
```diff
@@ -597,3 +604,55 @@ func (ci *ConnectionIterator) roundRobin(addresses []Address) {
...
+		for _, address := range attempt.Addresses {
+			if !ci.isDownHost(address) {
+				return false
+			}
+		}
+		return true
+	})
+}
+
+func (ci *ConnectionIterator) isDownHost(address Address) bool {
```
Comment (ejannett) [inline internal/driver/network/naming/iterator.go]: add missing doc

File: internal/driver/network/session/nsprotocol.go [hunk 9]
```diff
@@ -200,6 +211,44 @@ func (ns *networkSession) transportConnect(ctx context.Context, address transpor
...
+		return false
+	}
+
+	if errors.Is(err, syscall.EHOSTDOWN) ||
+		errors.Is(err, syscall.EHOSTUNREACH) ||
+		errors.Is(err, syscall.ENETUNREACH) {
+		return true
+	}
+
+	var timeoutCause common.CtxTimeoutCauseError
```
Comment (ejannett) [inline internal/driver/network/session/nsprotocol.go]: I ma confused by the meaning  of "down" here.  host down means a host that can be resolved , for which you have a proper network route but connection is impossible because of host is not responding. 

If so . some of the cases we treat here are not "host down" . 
EHOSTUNREACH / ENETUNREACH : means that you have no network route to the host. You do not know how to get there. 
Any timeout we have can be multiple root causes:  too short timeout configured ,  to current host being slow , network bandwidth being full and last but not least , client being to far away from server . 

It seems tricky to me to group all this cases  under the "host down" case. especially, when keeping in mind  that in most of the cases, the true issue is that user gave us a  bad  timeout value  and in such case that should be not our responsibility to blacklist an host  but to raise the issue back to the user asap.

File: internal/driver/network/session/nsprotocol_test.go [hunk 10]
```diff
@@ -281,6 +289,77 @@ func TestTransportConnect(t *testing.T) {
 	}
 }
 
+func TestIsDownHostError(t *testing.T) {
```
Comment (ejannett) [inline internal/driver/network/session/nsprotocol_test.go]: add doc

File: internal/driver/network/naming/down_hosts_cache.go (hunk 3 above)
Comment (ejannett) [reply internal/driver/network/naming/down_hosts_cache.go]: You store reference to string not the string or yo can at least store always the same value. no point to trigger an allocation even if that's small

**Review by feroal — APPROVED** [review]

File: internal/driver/network/naming/iterator.go [hunk 11]
```diff
@@ -233,6 +236,7 @@ func (ci *ConnectionIterator) Reset() {
 			ci.descAttempts[i].Description.IsLoadBalanceEnabled() {
 			ci.shuffleAddresses(ci.descAttempts[i].Addresses)
 		}
+		ci.reorderAddressesByDownHostStatus(ci.descAttempts[i].Addresses)
```
Comment (ashichau) [inline internal/driver/network/naming/iterator.go]: Reset() reorders addresses within each description, but it does not reorder ci.descAttempts. If a description becomes fully down after the iterator is created, Reset() can still try that description before a healthy description. Should Reset() also call reorderDescriptionsByDownHostStatus(ci.descAttempts)? 
I mean say ci.reorderAddressesByDownHostStatus(ci.descAttempts[i].Addresses)
It reorders only the addresses inside each description.
My suggested additional line would be after the loop:
for i := range ci.descAttempts {
    // reset and reorder each description's addresses
}

ci.reorderDescriptionsByDownHostStatus(ci.descAttempts)
Please add a test with two descriptions.

File: internal/driver/network/naming/down_hosts_cache.go (hunk 3 above)
Comment (raihan687) [reply internal/driver/network/naming/down_hosts_cache.go]: struct{}{} is a zero-sized value in Go, so Put(key, struct{}{}) does not allocate memory per call and does not store any value/reference in the cache entry.  
Using either Put(key, key) or a shared constant string would avoid copying string bytes, but the cache would still store an additional string header (reference and length) as the value for every entry. Since we only need membership, struct{}{} avoids that unused storage entirely

File: internal/driver/network/naming/iterator.go (hunk 7 above)
Comment (raihan687) [reply internal/driver/network/naming/iterator.go]: done thanks

File: internal/driver/network/naming/iterator.go (hunk 8 above)
Comment (raihan687) [reply internal/driver/network/naming/iterator.go]: done thanks

File: internal/driver/network/session/nsprotocol.go (hunk 9 above)
Comment (raihan687) [reply internal/driver/network/session/nsprotocol.go]: “Down” here is a temporary reachability hint, not a permanent blacklist. For EHOSTUNREACH and ENETUNREACH, we also want to deprioritise the node: another resolved IP may still be reachable through a different route/VIP.  
The node is not removed—it remains in the attempt list and is tried later in the queue as a fallback.

File: internal/driver/network/session/nsprotocol_test.go (hunk 10 above)
Comment (raihan687) [reply internal/driver/network/session/nsprotocol_test.go]: done thanks

File: internal/driver/network/naming/iterator.go (hunk 11 above)
Comment (raihan687) [reply internal/driver/network/naming/iterator.go]: Thanks for pointing out the edge case, I've added the test and updated the code as well.

File: internal/driver/network/session/nsprotocol.go (hunk 6 above)
Comment (raihan687) [reply internal/driver/network/session/nsprotocol.go]: Hi Ashish, I updated the code in the latest commit.
Thanks for pointing out.

File: internal/driver/network/session/nsprotocol.go [hunk 12]
```diff
@@ -200,6 +211,44 @@ func (ns *networkSession) transportConnect(ctx context.Context, address transpor
...
+	if errors.As(err, &timeoutCause) {
+		return true
+	}
+
+	var sqlErr oracleErrors.SQLError
```
Comment (ashichau) [inline internal/driver/network/session/nsprotocol.go]: Could CtxTimeout also represent a DNS resolution timeout? resolveAddress() keeps the original hostname when DNS resolution fails, and tcpadapter.go converts a timed-out *net.OpError into oracleErrors.CtxTimeout before this classifier runs. The original *net.DNSError is then lost, so this code may cache a DNS failure as a down endpoint, although there is a clear check where the DNS failures should not be cached even PR description say that. Could we preserve/check the DNS error before converting the timeout,

Comment (raihan687) [reply internal/driver/network/session/nsprotocol.go]: Good catch. DNS errors are now preserved before generic timeout conversion, so a DNS timeout remains identifiable and is excluded from the down-host cache. Added tests for error preservation and cache exclusion. Connection attempts and failover continue as before; DNS timeouts now retain the underlying DNS error instead of being reported as a generic driver timeout.
